### ISSUE TITLE
GLsizeiptr/GLintptr type change, depending on khrplatform

### ISSUE THREAD

**Issue by Dav1dde** [KhronosGroup/OpenGL-Registry#184]
Followup for: https://github.com/KhronosGroup/OpenGL-Registry/issues/162

1) Due to the change the GL API depends on `khrplatform.h` now which was previously not the case, this seems like a breaking change that may affect multiple users of the specification. This may also affect users that are using `glcorearb.h` directly (including OS package maintainers?).

2) The typedefinitions have a lot of duplicates now `<type requires="khrplatform">typedef khronos_intptr_t <name>GLintptr</name>;</type>`: `<type api="gles1" requires="khrplatform">typedef khronos_intptr_t <name>GLintptr</name>;</type>`, `<type api="gles2" requires="khrplatform">typedef khronos_intptr_t <name>GLintptr</name>;</type>` etc.

3) Extension types were not changed: `<type requires="stddef">typedef ptrdiff_t <name>GLintptrARB</name>;</type>`

Background information (shouldn't affect your decision on keeping this change):
[glad](https://github.com/Dav1dde/glad) an application to generate loaders based on these specifiactions broke, mainly due to a bug that was triggered by this and not the specification change itself. That's how I noticed the change (multiple users reported the generator breaking over night). My fix was to wrap the typedefs with a macro that checks if `khrplatform.h` was previously included, this works for glad but would not be enough for the specs.

I am not sure to fix this "better", it might be possible using macros, like guarding the type definition.

**Comment by oddhack** [comment]
Thanks for the field experience. I was afraid we'd run into problems from this change and that's being borne out :-(

1. The dependency on khrplatform.h is a new requirement. I was worried about this, but people in the OpenGL working group felt it was more important to have consistency between the two headers and decided to accept the change. I did add some text to the GL index page pointing to the location of khrplatform.h, and presumably Mesa and other packagers of these headers will pick up on that, but we don't really know who's using the XML and headers for what purposes so it wasn't clear how to be more proactive than that.

2. The duplication can probably be reduced safely now. I'll take a look at that.

3. This was an oversight by the author of the MR. The definition of the ARB and corresponding core types should be consistent. I'll also take a look at that.

**Comment by c0d1f1ed** [comment]
@Dav1dde thanks for reporting this and sorry about the trouble. It sounds like my change merely exposed a prior bug? Glad you were able to easily fix it.

The dependency on khrplatform.h is intentional as it fixes and ensures consistency between OpenGL ES and desktop OpenGL.

As far as I know the XML files are a Khronos implementation detail for generating the headers. Previously .spec files were used for this. So in theory one shouldn't depend on them, or breaking changes are to be expected. Someone correct me if I'm wrong.

So basically the majority of people affected by this change merely have to ensure khrplatform.h is available. The rest is probably prior bugs and implementation dependencies. If I'm not mistaken the only unexpected issue is the extension types were not changed at the same time?

That's my mistake. I did notice them but expected we could independently update them afterwards. @oddhack thanks for looking into it. Let me know if I can lend a hand.

**Comment by Dav1dde** [comment]
It was a somewhat unfortunate combination of a bug and the implicit assumption `khorons_*` types are not used for an OpenGL context.

The change definitely makes sense, I am just debating now if I took the right approach of always using the latest upstream specification, I thought it's a good idea simply because the latest changes are always immediately available, now this affects a lot of people. Something I have to consider.

**Comment by oddhack** [comment]
@c0d1f1ed gl.xml is the normative source of the headers and used for a variety of purposes by outside projects, so it's not just a detail. But it's impossible to know how all those projects are using it, and if they do their own processing of the XML rather than derive from the scripts we supply, or if they run into underdocumented portions of the schema or similar, they can expose problems.

@Dav1dde it is probably not robust to use the HEAD revision of gl.xml all the time. In general the only changes outside consumers are likely to really care about is when new extensions are added, but there's no good way to separate those commits out from other changes. Someday if I have spare time, I may transition over to the XML schema and scripts used by Vulkan, which has evolved a good deal beyond where the GL XML is. That would be a breaking change so would come with notice. But unlikely to happen anytime soon.

**Comment by oddhack** [comment]
@c0d1f1ed @Dav1dde please evaluate #186.
